ctapipe/tools/stage1.py

```diff
@@ -0,0 +1,868 @@
+            # we can write the configuration data.
+            if event.count == 0:
+                tel_list_transform = create_tel_id_to_tel_index_transform(
+                    event.inst.subarray
+                )
+                writer.add_column_transform(
+                    table_name="dl1/event/subarray/trigger",
+                    col_name="tels_with_trigger",
+                    transform=tel_list_transform,
+                )
+
+                self._write_simulation_configuration(writer, event)
+                self._write_instrument_configuration(event.inst.subarray)
+                is_initialized = True
+
+            # write the subarray tables
+            writer.write(
+                table_name="dl1/event/subarray/mc_shower",
+                containers=[event_index, event.mc],
+            )
+            writer.write(
+                table_name="dl1/event/subarray/trigger",
+                containers=[event_index, event.trig],
+            )
+            # write the telescope tables
+            self._write_telescope_event(writer, event, tel_index)
+
+        if is_initialized is False:
+            raise ValueError(f"No events found in file: {self.event_source.input_url}")
+
+    def _write_telescope_event(self, writer, event, tel_index):
+        """
+        add entries to the event/telescope tables for each telescope in a single
+        event
+        """
+
+        # write the telescope tables
+        for tel_id, data in event.dl1.tel.items():
+
+            data.prefix = ""  # don't want a prefix for this container
+            telescope = event.inst.subarray.tel[tel_id]
+            tel_type = str(telescope)
+            tel_index.tel_id = np.int16(tel_id)
+            tel_index.tel_type_id = tel_type_string_to_int(tel_type)
+            table_name = (
+                f"tel_{tel_id:03d}" if self.split_datasets_by == "tel_id" else tel_type
+            )
+
+            extra = ExtraImageContainer(
+                mc_photo_electron_image=event.mc.tel[tel_id].photo_electron_image,
+                selected_gain_channel=event.r1.tel[tel_id].selected_gain_channel,
+                image_mask=None,  # added later, if computed only
+            )
+
+            if self.write_parameters:
+
+                image_mask, params = self._parameterize_image(
+                    event.inst.subarray, data, tel_id=tel_id
+                )
+
+                self.log.debug("params: %s", params.as_dict(recursive=True))
+                self.log.debug("Writing! ******")
+
+                containers_to_write = [
+                    tel_index,
+                    params.hillas,
+                    params.timing,
+                    params.leakage,
+                    params.concentration,
+                    params.morphology,
+                    params.intensity,
+                ]
+
+                # currently the HDF5TableWriter has problems if the first event
```

Comment:
Is there then any problem for specifying them in the fields straightaway? Otherwise, could we for the time being not use nan but some conventional default error values? Having in the output the events for which the parameterisation failed may be useful. Another "ugly" alternative: we could just make sure that the first event has no nans, but keep all subsequent ones.

Reply:
Acutally, that was the original intent (to write all events) ,but due to how things work currently, it's not possible - at some point we changed many of the "Default values" in the containers to either None or np.nan, which makes it impossible to guess the data type if the info is not filled in (and is nan is not supported by pytables).  Since the schema is generated by the info in the first event, it fails if the first event is not parameterizable.  In #1224  I suggested a fix for that, but I think for now we have to wait for that fix, and then in the future, we can store all events.    It's not really clear what is a good "junk" value for all parameters, however, particularly for integers.

Reply:
This ja a hot topic currently nearly everywhere in data science. Newer file formats and libraries start supporting a general missing value indicator.

E.g. pandas since 1.0 and Apache arrow

Reply:
> and is nan is not supported by pytables?

What do you mean with that? pytables is perfectly able to store and read nans and will correctly deduce a type of float for nans:

```
In [10]: f = tables.open_file('/tmp/test.hdf5', 'w')                                                                                          

In [11]: f.create_array('/', 'all_nan', np.full((2, 1855), np.nan))                                                                           
Out[11]: 
/all_nan (Array(2, 1855)) ''
  atom := Float64Atom(shape=(), dflt=0.0)
  maindim := 0
  flavor := 'numpy'
  byteorder := 'little'
  chunkshape := None
```

Reply:
What is the error here, if nan is encountered? Why can we not write an event with nan in the hillas container?